src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -201,6 +194,47 @@ public void StartService()
                 throw;
             }
         }
+
+        private void Release(Guid agentId, ITestAgent agent)
+        {
+            if (_agentStore.IsAgentProcessActive(agentId, out var process))
+            {
+                try
+                {
+                    log.Debug("Stopping remote agent");
+                    agent.Stop();
+                }
+                catch (SocketException ex)
+                {
+                    int? exitCode;
+                    try
+                    {
+                        exitCode = process.ExitCode;
+                    }
+                    catch (NotSupportedException)
+                    {
+                        exitCode = null;
+                    }
+
+                    if (exitCode == 0)
+                    {
+                        log.Warning("Agent connection was forcibly closed. Exit code was 0, so agent shutdown OK");
+                    }
+                    else
+                    {
+                        var stopError = $"Agent connection was forcibly closed. Exit code was {exitCode?.ToString() ?? "unknown"}. {Environment.NewLine}{ExceptionHelper.BuildMessageAndStackTrace(ex)}";
+                        log.Error(stopError);
+                        throw new Exception(stopError, ex);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    var stopError = "Failed to stop the remote agent." + Environment.NewLine + ExceptionHelper.BuildMessageAndStackTrace(ex);
+                    log.Error(stopError);
+                    throw new Exception(stopError, ex);
```

Comment:
Throwing bare Exception is a smell, but so is catching bare Exception and we're already doing that here and in the calling method. Figured there was no reason to introduce a new exception type.